File: python/kvs.py
import collections
import json
import copy

KVS_KEY_TERMINATOR = '='
KVS_VALUE_TERMINATOR = ';'
KVS_STRUCT_START = '['
KVS_STRUCT_END = ']'
KVS_META_DATA = '~'

KVS_STATE_KEY = 0
KVS_STATE_VALUE = 1
KVS_STATE_META = 2


class INT(object):
    encoding = 'UTF-8'

    def __init__(self, init_value):
        self.value = init_value


class KVS(object):
    encoding = 'UTF-8'

    def __init__(self,  kvs=None, index=None):
        self.kvs = collections.OrderedDict()
        self.index = INT(0)
        self.depth = INT(1)
        self.default_key = INT(0)
        if isinstance(kvs, str):
            if index is None:
                self.index.value = 0
                self.__parse_kvs(kvs, self.index)
            else:
                self.__parse_kvs(kvs, index)
        elif isinstance(kvs, KVS):
            self.kvs = kvs.kvs

    def __dict_to_kvs(self, kvs, pretty, wrap):
        return_string = ''
        if wrap is not None:
            return_string += wrap
            return_string += '['
            if pretty is True:
                return_string += "\n\t"
        for key, value in kvs.items():
            if pretty is True:
                return_string += "\n"
                for i in range(0, self.depth.value):
                    return_string += "\t\t"
            return_string += key
            if isinstance(value, KVS):
                return_string += '['
                if pretty is True:
                    for i in range(0, self.depth.value):
                        return_string += "\t"
                    self.depth.value += 1
                return_string += self.__dict_to_kvs(value.kvs, pretty, None)
                if pretty is True:
                    return_string += "\n"
                    for i in range(0, self.depth.value):
                        return_string += "\t\t"
                    self.depth.value -= 1
                return_string += ']'
            else:
                return_string += '='
                if ';' in value:
                    new_string = ''
                    for c in value:
                        if c is ';':
                            new_string += ';'
                        new_string += c
                    return_string += new_string
                else:
                    return_string += value
                return_string += ';'
                if pretty is True:
                    return_string += "\t\t"
        if wrap is not None:
            return_string += ']'
        return return_string

    def __dump(self, obj):
        newobj = copy.deepcopy(obj)
        for key, value in newobj.items():
            if isinstance(value, KVS):
                tmpObj = self.__dump(newobj[key].kvs)
                newobj[key] = tmpObj
        return newobj

    def __parse_kvs(self, kvsString, index):
        array_counter = 0
        kvs_key_string = ''
        kvs_value_string = ''
        kvs_state = KVS_STATE_KEY
        while index.value < len(kvsString):
            currentChar = kvsString[index.value]
            if currentChar is KVS_KEY_TERMINATOR:
                if kvs_state == KVS_STATE_KEY:
                    kvs_state = KVS_STATE_VALUE
                elif kvs_state == KVS_STATE_META:
                    kvs_state = KVS_STATE_VALUE
                elif kvs_state == KVS_STATE_VALUE:
                    kvs_value_string += currentChar

            elif currentChar is KVS_META_DATA:
                kvs_state = KVS_STATE_META
            elif currentChar is KVS_VALUE_TERMINATOR:
                if index.value+1 < len(kvsString):
                    if currentChar is kvsString[index.value+1]:
                        index.value += 1
                        kvs_value_string += currentChar
                    else:
                        kvs_state = KVS_STATE_KEY
                        if kvs_key_string.strip() is '':
                            kvs_key_string = str(array_counter)
                            array_counter += 1
                        self.kvs[kvs_key_string.strip()] = kvs_value_string
                        kvs_key_string = ''
                        kvs_value_string = ''
                else:
                    kvs_state = KVS_STATE_KEY
                    if kvs_key_string.strip() is '':
                        kvs_key_string = str(array_counter)
                        array_counter += 1
                    self.kvs[kvs_key_string.strip()] = kvs_value_string
                    kvs_key_string = ''
                    kvs_value_string = ''
            elif currentChar is KVS_STRUCT_START:
                kvs_state = KVS_STATE_KEY
                index.value+=1
                if kvs_key_string.strip() is '':
                    kvs_key_string = str(array_counter)
                    array_counter += 1
                self.kvs[kvs_key_string.strip()] = KVS(kvsString, index)
                kvs_key_string = ''
                kvs_value_string = ''
            elif currentChar is KVS_STRUCT_END:
                return
            else:
                if kvs_state == KVS_STATE_KEY:
                    kvs_key_string += currentChar
                elif kvs_state == KVS_STATE_VALUE:
                    kvs_value_string += currentChar
            index.value += 1

    def add(self, value):
        while True:
            if str(self.default_key.value) in self.kvs.keys():
                self.default_key.value += 1
            else:
                self.kvs[str(self.default_key.value)] = value
                self.default_key.value += 1
                return

    def clear(self):
        self.kvs.clear()

    def exists(self, *argv):
        activeObject = copy.deepcopy(self.kvs)
        for arg in argv:
            if arg in activeObject.keys():
                if isinstance(activeObject[arg], KVS):
                    activeObject = activeObject[arg].kvs
                else:
                    return True
            else:
                return False

    def fromString(self, kvsString):
        self.__parse_kvs(kvsString, self.index)

    def get(self, *argv):
        activeObject = copy.deepcopy(self.kvs)
        for arg in argv:
            if arg in activeObject.keys():
                if isinstance(activeObject[arg], KVS):
                    activeObject = activeObject[arg].kvs
                else:
                    return activeObject[arg]
            else:
                return ""
        return ""

    def getFirstKey(self):
        if self.kvs.keys():
            return list(self.kvs.keys())[0]
        else:
            return ""

    def getFirstValue(self):
        if self.kvs.values():
            return list(self.kvs.values())[0]
        else:
            return ""

    def getKVS(self, *argv):
        activeObject = self
        for arg in argv:
            if arg in activeObject.kvs.keys():
                if isinstance(activeObject.kvs[arg], KVS):
                    activeObject = activeObject.kvs[arg]
                else:
                    return ""
            else:
                activeObject.kvs[arg] = KVS()
                activeObject = activeObject.kvs[arg]
        return activeObject

    def getNative(self, *argv):
        activeObject = self
        for arg in argv:
            if arg in activeObject.kvs.keys():
                if isinstance(activeObject.kvs[arg], KVS):
                    activeObject = activeObject.kvs[arg]
                else:
                    return ""
            else:
                activeObject.kvs[arg] = KVS()
                activeObject = activeObject.kvs[arg]
        return activeObject.kvs

    def keys(self):
        return list(self.kvs.keys())

    def merge(self, kvs, sub=None):
        baseKVS = self
        if sub is not None:
            baseKVS = sub
        for key in kvs.kvs.keys():
            if isinstance(kvs.kvs[key], KVS):
                self.merge(kvs.kvs[key], baseKVS.kvs[key])
            else:
                baseKVS.kvs[key] = kvs.kvs[key]

    def remove(self, *argv):
        activeObject = self.kvs
        for arg in argv:
            if arg in activeObject.keys():
                if isinstance(activeObject[arg], KVS):
                    activeObject = activeObject[arg].kvs
                else:
                    activeObject.pop(arg, None)

    def set(self, value, *argv):
        activeObject = self.kvs
        for i in range(0, len(argv)):
            if argv[i] in activeObject.keys():
                if isinstance(activeObject[argv[i]], KVS):
                    if i < len(argv)-1:
                        activeObject = activeObject[argv[i]].kvs
            else:
                if i < len(argv) - 1:
                    activeObject[argv[i]] = KVS()
                    activeObject = activeObject[argv[i]].kvs
        activeObject[argv[i]] = value

    def toString(self, pretty=False, wrap=None):
        return self.__dict_to_kvs(self.kvs, pretty, wrap)

    def values(self):
        return list(self.kvs.values())

    def debug(self):
        print(json.dumps(self.__dump(self.kvs), indent=4, default=str))

File: cpp/KVS.cpp
/*
 * KVS.cpp
 *
 *  Created on: 13 Feb 2019
 *      Author: hansen
 */

#include "KVS.h"

#define KVS_KEY_TERMINATOR    '='
#define KVS_VALUE_TERMINATOR  ';'
#define KVS_STRUCT_START      '['
#define KVS_STRUCT_END        ']'
#define KVS_META_DATA         '~'

#define KVS_STATE_KEY    0
#define KVS_STATE_VALUE  1
#define KVS_STATE_META   2

KVS::KVS(){ kvs_index = 0; }
KVS::KVS(const KVS& kvs) { this->dictionary = kvs.dictionary; kvs_index = 0;}
KVS::KVS(string kvs, int index){
	kvs_index = index;
	fromString(kvs, kvs_index);
}

KVS::~KVS(){}

void KVS::fromString(string kvs, int &kvs_index){
	int array_counter = 0;
	std::string key_string = "";
	std::string value_string = "";
	int kvs_state = KVS_STATE_KEY;
	while(kvs_index < kvs.length()){
		char currentChar = kvs.at(kvs_index);
		if(currentChar == KVS_KEY_TERMINATOR){
			if(kvs_state == KVS_STATE_KEY)
				kvs_state = KVS_STATE_VALUE;
			else if(kvs_state == KVS_STATE_META)
				kvs_state = KVS_STATE_VALUE;
			else if(kvs_state == KVS_STATE_VALUE)
				value_string += currentChar;
		}else if(currentChar == KVS_META_DATA){
			kvs_state = KVS_STATE_META;
		}else if(currentChar == KVS_VALUE_TERMINATOR){
			if(kvs_index+1 < kvs.length()){
				if(currentChar == kvs.at(kvs_index+1)){
					kvs_index++;
					value_string += currentChar;
				}else{
					kvs_state = KVS_STATE_KEY;
					trim(key_string);
					if(key_string.empty())
						key_string = to_string(array_counter++);
					dictionary[key_string] = KVS_Value(KVS_VALUE_STRING, (void*)value_string.c_str(), value_string.length());
					key_string = "";
					value_string = "";
				}
			}else{
				kvs_state = KVS_STATE_KEY;
				trim(key_string);
				if(key_string.empty())
					key_string = to_string(array_counter++);
				dictionary[key_string] = KVS_Value(KVS_VALUE_STRING, (void*)value_string.c_str(), value_string.length());
				key_string = "";
				value_string = "";
			}
		}else if(currentChar == KVS_STRUCT_START){
			kvs_state = KVS_STATE_KEY;
			kvs_index++;
			trim(key_string);
			if(key_string.empty())
				key_string = to_string(array_counter++);
			dictionary[key_string] = KVS_Value(KVS_VALUE_STRUCT, (void*)new KVS(kvs, kvs_index), sizeof(KVS));
			kvs_index += ((KVS*)dictionary[key_string].value)->kvs_index - kvs_index;
			key_string = "";
			value_string = "";
		}else if(currentChar == KVS_STRUCT_END){
			if(kvs_state == KVS_STATE_KEY)
				return;
		}else{
			if(kvs_state == KVS_STATE_KEY)
				key_string += currentChar;
			else if(kvs_state == KVS_STATE_VALUE)
				value_string += currentChar;
		}

		kvs_index++;
	}
	for(map<string,KVS_Value>::iterator it = dictionary.begin(); it != dictionary.end(); ++it){
			keys.push_back(it->first);
			values.push_back(it->second);
	}

}

string KVS::__dict_to_kvs(map<string,KVS_Value> active){
	string returnString = "";
	for(map<string,KVS_Value>::iterator it = active.begin(); it != active.end(); ++it) {
		if(it->second.type == KVS_VALUE_STRING){
			returnString.append(it->first);
			returnString.append("=");
			string value = (char*)it->second.value;
			value = do_replace(value, ";", ";;");
			returnString.append(value);
			returnString.append(";");
		}else{
			returnString.append(it->first);
			returnString.append("[");
			returnString.append(__dict_to_kvs(((KVS*)it->second.value)->dictionary));
			returnString.append("]");
		}

	}
	return returnString;
}

string KVS::toString(){
	return __dict_to_kvs(dictionary);
}

void KVS::__garbage_collect(map<string,KVS_Value> active){
	for(map<string,KVS_Value>::iterator it = active.begin(); it != active.end(); ++it) {
		if(it->second.type == KVS_VALUE_STRING)
			classes.push_back(it->second.value);
		else{
			classes.push_back(it->second.value);
			__garbage_collect(((KVS*)active[it->first].value)->dictionary);
		}
	}
}

void KVS::Add(KVS_Value value){
	while(true){
		if(find(keys.begin(), keys.end(), to_string(autokey)) != keys.end()){
			autokey++;
		}else{
			dictionary[to_string(autokey++)] = value;
			return;
		}
	}
}

void KVS::Clear(){
	__garbage_collect(dictionary);
	for(int i = 0; i < classes.size(); i++)
		free(classes.at(i));
	dictionary.clear();
}

void KVS::Debug(map<string,KVS_Value> active){
	for(map<string,KVS_Value>::iterator it = active.begin(); it != active.end(); ++it) {
		if(it->second.type == KVS_VALUE_STRING)
			cout << it->first << " = " << (char*)it->second.value << "\n";
		else{
			cout << it->first << "\n";
			Debug(((KVS*)active[it->first].value)->dictionary);
		}
	}
	cout << "\n";
}

bool KVS::Exists(int count, ...){
	bool returnBool = false;
	va_list args;
	va_start(args, count);
	vector<string> active = keys;
	for(int i = 0; i < count; i++){
		string currentKey = string(va_arg(args, char*));
		cout << i << " " << currentKey << " " << "\n";
//		if(find(active.begin(), active.end(), string(currentKey)) != keys.end()){
			for(int j = 0; j < active.size(); j++)
				cout << "LOOKING FOR " << currentKey << " : " << active.at(j) << "\n";
//			if(dictionary[string(currentKey)].type == KVS_VALUE_STRUCT){
//				cout << "FOUND\n";
//				active = ((KVS*)dictionary[string(currentKey)].value)->keys;
//			}
//
//		}else{
//			cout << "NOT FOUND\n";
//			returnBool = false;
//		}
		for(int j = 0; j < active.size(); j++){
			if(currentKey == active.at(j)){
				if(dictionary[currentKey].type == KVS_VALUE_STRUCT){
					active = ((KVS*)dictionary[currentKey].value)->keys;
					returnBool = true;
					break;
				}else if(i == count-1){
					returnBool = true;
					break;
				}
			}else{
				returnBool = false;
			}
		}
	}
	va_end(args);
	return returnBool;
}
